File: Leetcode_Problems/0572.py
# LeetCode 0572 - Subtree of Another Tree
# Easy - Trees
# Given the roots of two binary trees root and subRoot, return true if there is a subtree of root with the same structure and node values of subRoot and false otherwise.
#
# A subtree of a binary tree tree is a tree that consists of a node in tree and all of this node's descendants. The tree tree could also be considered as a subtree of itself.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Recursive Depth-First Search Solution
# Complexities:
# Time : O(m * n) - m = number of nodes in root, n = number of nodes in subRoot
# Space: O(m + n)
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot: return True
        if not root: return False
        return True if self.sameTree(root, subRoot) else (self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot))

    # s is root, t is subRoot
    def sameTree(self, s, t):
        if not s and not t:
            return True
        if s and t and s.val == t.val:
            return self.sameTree(s.left, t.left) and self.sameTree(s.right, t.right)
        return False

File: Leetcode_Problems/0572.go
// LeetCode 0572 - Subtree of Another Tree
// Easy - Trees
// Given the roots of two binary trees root and subRoot, return true if there is a subtree of root with the same structure and node values of subRoot and false otherwise.
// A subtree of a binary tree tree is a tree that consists of a node in tree and all of this node's descendants. The tree tree could also be considered as a subtree of itself.

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Recursive Depth-First Search Solution
// Complexities:
// Time : O(m * n)
// Space: O(m + n)
func helperSameTree(root *TreeNode, subRoot * TreeNode) bool {
    if root == nil && subRoot == nil { return true }
    if root != nil && subRoot != nil && root.Val == subRoot.Val {
        return helperSameTree(root.Left, subRoot.Left) && helperSameTree(root.Right, subRoot.Right)
    }
    return false
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
    if subRoot == nil { return true }
    if root == nil { return false }
    if helperSameTree(root, subRoot) { return true }
    return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}
